src/storage.py

```diff
@@ -0,0 +1,283 @@
+
+    def _on_storage_detaching(self, event: StorageDetachingEvent):
+        """Unified storage detaching handler."""
+        # check if the detaching device is being used as or with an OSD.
+        osd_num = self._get_osd_num(event.storage.location.as_posix(), event.storage.name)
+
+        if osd_num:
+            with sunbeam_guard.guard(self.charm, self.name):
+                try:
+                    self.remove_osd(osd_num)
+                except CalledProcessError as e:
+                    if self._is_safety_failure(e.stderr):
+                        warning = f"Storage {event.storage.full_id} detached, provide replacement for osd.{osd_num}."
+                        logger.warning(warning)
+                        # clean records since juju will deprovision device.
+                        self.remove_osd(osd_num, force=True)
+                        raise sunbeam_guard.BlockedExceptionError(warning)
+
+    # helper functions
+
+    def _fetch_filtered_storages(self, directives: list) -> list:
+        """Provides a filtered list of attached storage devices."""
+        filtered = []
+        for device in self.juju_storage_list():
+            if device.split("/")[0] in directives:
+                filtered.append(device)
+
+        return filtered
+
+    def _is_safety_failure(self, err: str) -> bool:
+        """Checks if the subprocess error is caused by safety check."""
+        return "need at least 3 OSDs" in err
+
+    def _run(self, cmd: list) -> str:
+        """Wrapper around subprocess run for storage commands."""
+        process = run(cmd, capture_output=True, text=True, check=True, timeout=180)
+        logger.debug(f"Command {' '.join(cmd)} finished; Output: {process.stdout}")
+        return process.stdout
+
+    def _enroll_with_db(self, disk: list, db: list):
+        """Checks if sufficient devices are available to be enrolled into OSDs."""
+        enrollment_count = min(len(disk), len(db))
+        for i in range(enrollment_count):
+            try:
+                microceph.add_osd_cmd(spec=disk[i], db_dev=db[i])
+                # store configured devices.
+                self._save_osd_data(disk=disk[i], db=db[i])
+            except CalledProcessError as e:
+                logger.error(e.stderr)
+
+    def _enroll_disks_in_batch(self, disks: list):
+        """Adds requested Disks to Microceph and stored state."""
+        microceph.enroll_disks_as_osds(disks)
+        for disk in disks:
+            self._save_osd_data(disk)
+        logger.debug(f"Added {disks} as OSDs.")
+
+    def remove_osd(self, osd: int, force: bool = False):
+        """Removes OSD from MicroCeph and from stored state."""
+        try:
+            microceph.remove_disk_cmd(osd, force)
+            self._remove_osd_data(self._stored.osd_data[osd])
+        except CalledProcessError as e:
+            if force:
+                # clean stored state.
+                self._remove_osd_data(self._stored.osd_data[osd])
```

Comment:
ACK, This has been changed to remove only stale records. This jumping around is required because we do not want to remove the state records before MicroCeph DB records.